### PR TITLE
[tr064] reduce network load and improve XML handling

### LINKED ISSUE #9689: [tr064] Failed to get call list
Beside a couple of `TimeoutException`s I sporadically see messages like these:

```
2021-01-04 14:09:33.194 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processMissedCalls failed: Failed to get call list from URL https://<IP>:49443/calllist.lua?sid=611406051bc07a65
2021-01-04 14:09:33.199 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=missedCalls, getAction=GetCallList, dataType='string, parameter='7'}: failed to transform 'NewCallListURL'

2021-01-04 07:28:47.039 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processOutboundCalls failed: Failed to get call list from URL https://<IP>:49443/calllist.lua?sid=611406051bc07a65
2021-01-04 07:28:47.041 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=outboundCalls, getAction=GetCallList, dataType='string, parameter='7'}: failed to transform 'NewCallListURL'

2021-01-04 07:38:07.166 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processInboundCalls failed: Failed to get call list from URL https://<IP>:49443/calllist.lua?sid=611406051bc07a65
2021-01-04 07:38:07.169 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=inboundCalls, getAction=GetCallList, dataType='string, parameter='7'}: failed to transform 'NewCallListURL'

2021-01-04 07:38:01.926 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://<IP>:49443/tamcalllist.lua?sid=611406051bc07a65&tamindex=0
2021-01-04 07:38:01.928 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='0'}: failed to transform 'NewURL'
```

@J-N-K  Do you have an idea what is going on here?

ftr - `TimeoutException` - should we maybe increase the timeout or reduce log level?

```
2021-01-04 18:25:35.278 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=wanMaxUpstreamRate, getAction=GetCommonLinkProperties, dataType='ui4, parameter='null'}: java.util.concurrent.TimeoutException: Total timeout 2000 ms elapsed
```

### PR DESCRIPTION
**Pull request by J-N-K**
Fixes #9689

### CODE REVIEW

**Comment by cweitkamp** [comment]
Yesterday I installed this version for testing. The reported issue in #9689 did not occur again. Unfortunately the timeout is still there, but not that often - it looks like the new timeout is not taken into account:

```
2021-01-06 14:57:26.436 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=macOnline, getAction=GetSpecificHostEntry, dataType='boolean, parameter='72:2E:43:A4:44:9E'}: java.util.concurrent.TimeoutException: Total timeout 2000 ms elapsed
```

**Comment by J-N-K** [comment]
There's two different timeouts, one for SOAP requests and the other for all other HTTP requests. I have now adjusted bothto 5s.

**Comment by J-N-K** [comment]
I also implementing caching for the SOAP requests. Especially for call lists there are many requests which are the same.

File: bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java [hunk 1]
```diff
@@ -118,19 +122,46 @@ private Request prepareSOAPRequest(SCPDServiceType service, String soapAction, M
...
      * @return the SOAPMessage answer from the remote host
      * @throws Tr064CommunicationException if an error occurs during the request
      */
-    public synchronized SOAPMessage doSOAPRequest(SCPDServiceType service, String soapAction,
-            Map<String, String> arguments) throws Tr064CommunicationException {
+    public SOAPMessage doSOAPRequest(SOAPRequest soapRequest) throws Tr064CommunicationException {
         try {
-            Request request = prepareSOAPRequest(service, soapAction, arguments).timeout(SOAP_TIMEOUT,
-                    TimeUnit.MILLISECONDS);
+            SOAPMessage soapMessage = Objects.requireNonNull(soapMessageCache.putIfAbsentAndGet(soapRequest, () -> {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java:133]: How are you making sure that requests that modify remote state aren't getting cached? Are all requests getters?

File: bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java [hunk 2]
```diff
@@ -186,7 +216,9 @@ public void sendChannelCommandToDevice(Tr064ChannelConfig channelConfig, Command
                                     channelConfig.getChannelTypeDescription().getGetAction().getParameter().getName(),
                                     parameter);
                         }
-                        doSOAPRequest(service, channelTypeDescription.getSetAction().getName(), arguments);
+                        SOAPRequest soapRequest = new SOAPRequest(service,
+                                channelTypeDescription.getSetAction().getName(), arguments);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java:220]: @cpmeister This is the only place where set actions take place and it bypasses the cache by calling `doSOAPRequestUncached`

**Review by cpmeister — APPROVED** [review]

**Comment by cweitkamp** [comment]
Looks good so far. No messages in my log anymore for more than twelve hours. Thank you for looking into it.

File: bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java [hunk 3]
```diff
@@ -57,12 +58,16 @@
...
 public class SOAPConnector {
-    private static final int SOAP_TIMEOUT = 2000; // in ms
+    private static final int SOAP_TIMEOUT = 5; // in s
+    private static final int SOAP_MESSAGE_CACHE_EXPIRY_TIME = 2000; // in ms
     private final Logger logger = LoggerFactory.getLogger(SOAPConnector.class);
     private final HttpClient httpClient;
     private final String endpointBaseURL;
     private final SOAPValueConverter soapValueConverter;
 
+    private final ExpiringCacheMap<SOAPRequest, SOAPMessage> soapMessageCache = new ExpiringCacheMap<>(
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java:68]: Only a minor remark. Why not passing the time value directly like [here](https://github.com/openhab/openhab-addons/pull/9693/files#diff-5693a957fcf128a6a57fa9f5062b873f08f373818f4f1351bfe01705274af0aaR85) ? Or use `Duration.of()` to improve readability.

File: bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/util/Util.java [hunk 4]
```diff
@@ -67,6 +68,12 @@
 @NonNullByDefault
 public class Util {
     private static final Logger LOGGER = LoggerFactory.getLogger(Util.class);
+    private static final int HTTP_REQUEST_TIMEOUT = 5; // in s
+    private static final int XML_OBJECT_CACHE_EXPIRY_TIME = 3000; // in ms
+
+    // cache XML content for 5s
+    private static final ExpiringCacheMap<String, Object> XML_OBJECT_CACHE = new ExpiringCacheMap<>(
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/util/Util.java:75]: see last comment

**Review by cweitkamp — COMMENTED** [review]
LGTM. Thanks. And it is working fine. No more warnings since nearly two days.

File: bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java (hunk 3 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java:68]: The stateCache can probably be omitted. Sinc ethe SOAP request to fill it is cached anyway. WDYT?

Comment (cweitkamp) [reply bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java:68]: Yes, we probably could remove that one.

**Comment by cweitkamp** [comment]
I found another occurring message - is this handled differently?

```
2021-01-08 19:09:02.538 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://<IP>:49443/tamcalllist.lua?sid=1156132f00cbd862&tamindex=0
2021-01-08 19:09:02.545 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='0'}: failed to transform 'NewURL'
```

File: bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java (hunk 3 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/soap/SOAPConnector.java:68]: OTOH, it bypasses the construction of the SOAP request and does not add much complexity now that it is here. So I would prefer to leave it in place.

**Comment by J-N-K** [comment]
That request is not cached because there is only one request per answering machine. I wonder if the problem is that two requests are made in the same time (e.g. the request for the TAM list and a call list).

**Comment by cweitkamp** [comment]
But there are two channels available for each TAM: `tamEnable_n` and `tamNewMessages_n`.

**Comment by J-N-K** [comment]
True. But „enabled“ is from a regular (cached) soap request (like getting the URL) and only „newMessages“ is an additional HTTP call.

**Review by cweitkamp — APPROVED** [review]
LGTM. Thanks.

**Comment by sheilbronn** [comment]
BTW: I still get these messages on OpenHAB 3.2.0~M5 and Fritzbox software release 7.29 irregularly  :(

`2021-12-18 16:12:28.060 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://192.168.178.1:49443/tamcalllist.lua?sid=42b240695a650bbf&tamindex=0

2021-12-18 16:12:28.064 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='0'}: failed to transform 'NewURL'

2021-12-18 16:36:54.320 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://192.168.178.1:49443/tamcalllist.lua?sid=42b240695a650bbf&tamindex=0

2021-12-18 16:36:54.322 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='0'}: failed to transform 'NewURL'

2021-12-18 16:43:00.106 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://192.168.178.1:49443/tamcalllist.lua?sid=42b240695a650bbf&tamindex=0

2021-12-18 16:43:00.108 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='0'}: failed to transform 'NewURL'

2021-12-18 18:16:23.491 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://192.168.178.1:49443/tamcalllist.lua?sid=42b240695a650bbf&tamindex=0

2021-12-18 18:16:23.493 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='0'}: failed to transform 'NewURL'

2021-12-18 18:18:27.161 [INFO ] [064.internal.soap.SOAPValueConverter] - Postprocessor processTamListURL failed: Failed to get TAM list from URL https://192.168.178.1:49443/tamcalllist.lua?sid=42b240695a650bbf&tamindex=1

2021-12-18 18:18:27.162 [WARN ] [ng.tr064.internal.soap.SOAPConnector] - Failed to get Tr064ChannelConfig{channelType=tamNewMessages, getAction=GetMessageList, dataType='string, parameter='1'}: failed to transform 'NewURL'`

**Pull request merged**